File: firmware/src/clock.js
/**
 * Real Life Lonely T-Rex Game Firmware
 *
 * Copyright (C) 2017, 2018, Uri Shaked
 */

const assets = require('./assets');
const display = require('./display');

let updateTimer = null;

function displayString(s, x, y, charWidth, charHeight, spacing, prefix) {
  s.split('').forEach((c, i) => {
    if (c !== ' ') {
      display.writeChar(
        assets[(prefix || '') + c],
        charWidth,
        charHeight,
        y - i * (charHeight + spacing),
        x
      );
    }
  });
}

function zeroPad(val) {
  return (val < 10 ? '0' : '') + val.toString();
}

function displayTime(time) {
  return display.cls().then(() => {
    const top = 40;
    let hrs = zeroPad(time.getHours());
    let mins = zeroPad(time.getMinutes());
    display.fillMemory(0xff);
    displayString(hrs, top, 220, 48, 38, 10);
    display.writeChar(assets.colon, 48, 11, 220 - 74, top);
    displayString(mins, top, 220 - 130, 48, 38, 10);
    return display.displayFrame();
  });
}

function clockTask() {
  const date = new Date();
  const timeToUpdate =
    60000 - (date.getSeconds() * 1000 + date.getMilliseconds()) + 10;
  display.registerUpdate(() => displayTime(date));
  updateTimer = setTimeout(clockTask, timeToUpdate);
}

function stop() {
  if (updateTimer) {
    clearTimeout(updateTimer);
    updateTimer = null;
  }
}

function start() {
  stop();
  clockTask();
}

module.exports = {
  displayTime,
  start,
  stop
};

File: firmware/src/clock.py
"""
Real Life Lonely T-Rex Game Firmware

Copyright (C) 2017-2024, Uri Shaked
"""

import time
from game_assets import assets
from machine import Timer

## Hardware
from machine import Pin, SPI
from display import Display

spi = SPI(0, 2000000, sck=Pin(6), mosi=Pin(7))
cs = Pin(5)
dc = Pin(8)
rst = Pin(9)
busy = Pin(10)
display = Display(spi, cs, dc, rst, busy)
###

update_timer = Timer(-1)


def zero_pad(val):
    return f"{val:02}"


def display_time(tm: time.struct_time):
    hours = tm[3]
    minutes = tm[4]
    display.clear()
    x = 40
    display.draw_string(zero_pad(hours), x, 220)
    display.draw_string(":", x, 220 - 74)
    display.draw_string(zero_pad(minutes), x, 220 - 130)
    display.update()


def clock_task(timer: Timer = None):
    tm = time.localtime()
    seconds = tm[5]
    time_to_update = 60000 - seconds * 1000 + 10
    display_time(tm)

    timer.init(mode=Timer.ONE_SHOT, period=time_to_update, callback=clock_task)


def stop():
    global update_timer
    update_timer.deinit()


def start():
    global update_timer
    stop()
    clock_task(update_timer)


if __name__ == "__main__":
    start()
